File: laundry/wash_wallet_be/resources/js/types/service_package.ts
import {
    BaseFilters,
    BaseSortOptions,
    CustomerSubscription,
    Outlet,
    ServicePackageItem,
} from ".";

export interface ServicePackage {
    id: number;
    outletId: number | null;
    name: string;
    price: number;
    validityDays?: number | null;
    description?: string | null;
    isActive: boolean;
    isGlobal: boolean;
    createdAt: string;
    updatedAt: string;
    deletedAt?: string | null;
    outlet: Outlet;
    servicePackageItems: ServicePackageItem[];
    customerSubscriptions?: CustomerSubscription[];
    servicePackageItemsCount?: number;
    customerSubscriptionsCount?: number;
}

export interface ServicePackageFilters extends BaseFilters {
    outletId?: number | null;
    isActive?: boolean;
    minPrice?: number | null;
    maxPrice?: number | null;
    minValidityDays?: number | null;
    maxValidityDays?: number | null;
}

export interface ServicePackageSortOptions extends BaseSortOptions {
    column:
        | "name"
        | "price"
        | "validityDays"
        | "isActive"
        | "createdAt"
        | "updatedAt";
}

export interface ServicePackageFormData {
    name: string;
    description?: string | null;
    price: number;
    validityDays?: number | null;
    outletId?: number | null;
    isActive?: boolean | null;
    servicePackageItems: ServicePackageItemFormData[];
}

export interface ServicePackageItemFormData {
    laundryServiceId: number;
    quantity: number;
}

File: laundry/wash_wallet/packages/wash_wallet_domain/lib/src/entities/service_package.dart
import 'package:equatable/equatable.dart';

class ServicePackage extends Equatable {
  final int id;
  final int? outletId;
  final String name;
  final double price;
  final int? validityDays;
  final String? description;
  final bool isActive;
  final String? createdAt;
  final String? updatedAt;
  final String? deletedAt;
  final dynamic outlet;
  final List<dynamic> servicePackageItems;
  final List<dynamic> customerSubscriptions;
  final int servicePackageItemsCount;
  final int customerSubscriptionsCount;

  const ServicePackage({
    required this.id,
    this.outletId,
    required this.name,
    required this.price,
    this.validityDays,
    this.description,
    this.isActive = true,
    this.createdAt,
    this.updatedAt,
    this.deletedAt,
    this.outlet,
    this.servicePackageItems = const [],
    this.customerSubscriptions = const [],
    this.servicePackageItemsCount = 0,
    this.customerSubscriptionsCount = 0,
  });

  factory ServicePackage.fromModel(dynamic model) {
    return ServicePackage(
      id: model.id,
      outletId: model.outletId,
      name: model.name,
      price: model.price,
      validityDays: model.validityDays,
      description: model.description,
      isActive: model.isActive,
      createdAt: model.createdAt,
      updatedAt: model.updatedAt,
      deletedAt: model.deletedAt,
      outlet: model.outlet,
      servicePackageItems: model.servicePackageItems ?? [],
      customerSubscriptions: model.customerSubscriptions ?? [],
      servicePackageItemsCount: model.servicePackageItemsCount,
      customerSubscriptionsCount: model.customerSubscriptionsCount,
    );
  }

  @override
  List<Object?> get props => [
    id,
    outletId,
    name,
    price,
    validityDays,
    description,
    isActive,
    createdAt,
    updatedAt,
    deletedAt,
    outlet,
    servicePackageItems,
    customerSubscriptions,
    servicePackageItemsCount,
    customerSubscriptionsCount,
  ];
}
